src/main/groovy/io/fourfinanceit/pipeline/example/jobs/BuildAndPublish.groovy

```diff
@@ -22,29 +29,22 @@ import static io.fourfinanceit.pipeline.example.JenkinsVariable.SCM_POLLING_ENAB
  * @author Marek Kapowicki
  */
 class BuildAndPublish extends MicroserviceJobDefinition {
+    
     @Override
     void configure(Job job, MicroserviceProject project, JenkinsVariables jenkinsVariables) {
-        job.with {
-            setupEnvironmentVariables(job)
-            configure JobDslConfigureHacksUtil.injectAndMaskPasswords()
-            scm MicroserviceJobUtils.configureGit(project)
-            if (project.scmConfig.cronToPoll && jenkinsVariables.getBoolean(SCM_POLLING_ENABLED, true)) {
-                triggers {
-                    scm(project.scmConfig.cronToPoll)
-                }
-            }
-            addGradleBuildStep(job, project)
-            //addArchiveJunitAsBuildStep(job)
-            steps {
-                ifBuildStillSuccessful(delegate as StepContext) {
-                    gradle("publish --stacktrace ${projectVersionParameters()}")
-                }
-            }
-        }
+        (
+            setupEnvironmentVariables() &
+                maskPasswords() &
+                configureGit(project) &
+                use(onCron(project.scmConfig.cronToPoll))
+                    .when(scmPoolingEnabled(jenkinsVariables)) &
+                buildProject() &
+                publishArtifacts()
+        ).applyOn(job)
```

Comment:
Ok so there are few things with this. 
1. I prefer the notation with `()` since then the context of whole expression is clear, including the  applyOn invocation. I'm also almost certain that I've seen it working in internal 4Finance projects, although couldn't replicate it locally (maybe @szpak or @marekkapowicki could track the difference)
1. We've been working with this components for some time and no-one was puzzled by them, or at least no-one mentioned that. 
1. JobComponent is an Interface right now, so you can't make applyOn private + even if we move it to groovy, you don't want to disable `@CompileStatic ` on components creation. On addition, the fact of JobComponent being a functional interface right now does have few benefits
1. If someone would really want to use the fluent api, one can use the method as is (not a perfect solution, I agree).
1. I've just had an idea and I think that there is another possibility to solve this issue in much better way, however I couldn't make it working yet. I need few more experiments but the general idea is to be able to do something like:
```groovy
job << component +
    component + 
    anotherComponent
```
(& won't work, again due to parameters ordering)
It's working on a build, but not on jenkins - I need to investigate why.